services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -712,97 +891,176 @@ private Runnable getRunnableForDeferredVersionSwap() {
+                    parentStore,
+                    cluster,
+                    kafkaTopicName)) {
+                  continue;
+                }
 
-            // Switch to the target version in the completed non target regions
-            try {
-              rollForwardToTargetVersion(nonTargetRegionsCompleted, parentStore, targetVersion, cluster, repository);
-            } catch (Exception e) {
-              LOGGER.warn("Failed to roll forward for store: {} in version: {}", storeName, targetVersionNum, e);
+                emitMetricIfVersionSwapIfStalledForSequentialRollout(
+                    nextRegionToRollForward,
+                    parentStore,
+                    targetVersionNum,
+                    cluster);
 
-              int attemptedRetries = failedRollforwardRetryCountMap.compute(kafkaTopicName, (k, v) -> {
-                if (v == null) {
-                  return 1;
+                if (!didPostVersionSwapValidationsPass(
+                    parentStore,
+                    targetVersionNum,
+                    cluster,
+                    parentStore.getTargetSwapRegion(),
+                    kafkaTopicName)) {
+                  continue;
                 }
-                return v + 1;
-              });
-
-              if (attemptedRetries == MAX_ROLL_FORWARD_RETRY_LIMIT) {
-                deferredVersionSwapStats.recordDeferredVersionSwapFailedRollForwardSensor();
+              }
 
-                parentStore.updateVersionStatus(targetVersionNum, VersionStatus.PARTIALLY_ONLINE);
-                repository.updateStore(parentStore);
-                failedRollforwardRetryCountMap.remove(kafkaTopicName);
+              try {
                 LOGGER.info(
-                    "Updated parent version status to PARTIALLY_ONLINE for version: {} in store: {} after failing to roll forward in non target regions: {}",
-                    targetVersionNum,
+                    "Issuing roll forward for store: {} in region: {} for version: {}",
                     storeName,
-                    nonTargetRegionsCompleted);
+                    nextRegionToRollForward,
+                    targetVersionNum);
+                veniceParentHelixAdmin
+                    .rollForwardToFutureVersion(cluster, parentStore.getName(), nextRegionToRollForward);
+                storeWaitTimeCacheForSequentialRollout.invalidate(kafkaTopicName);
+
+                if (stalledVersionSwapSet.contains(parentStore.getName())) {
+                  stalledVersionSwapSet.remove(parentStore.getName());
+                }
 
-                // TODO cleanup the code below after parent VT is deprecated
-                LOGGER.info("Truncating kafka topic: {} after roll forward failed in 1+ regions", kafkaTopicName);
-                veniceParentHelixAdmin.truncateKafkaTopic(kafkaTopicName);
+                if (rolloutOrder.indexOf(nextRegionToRollForward) == rolloutOrder.size() - 1) {
```

Comment:
`indexOf` is O(n), rather check
rolloutOrder.get(rolloutOrder.size()-1).equals(nextRegionToRollForward)